### PR TITLE
[`core`] officially support SFT (Supervised Finetuning) 

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

This PR introduces `SFTTrainer` class. A handy and easy to use class to train your supervised fine-tuned model on instruction-based datasets. 
The API is easy to use, and also modular enough if you want to customize your training for advanced users.

You just need to pass a model id, optionally pass a `PeftConfig` to train adapters only. Pass also `from_pretrained` kwargs directly to `SFTTrainer` for advanced users, for example to load your model in 8bit mode.

The PR also introduces `ConstantLengthDataset`, a handy class to create instruction-based datasets. Just pass a tokenizer, a dataset, and a function to specify the formatting you want to have, and you should be good to go

Quickstart:
```python
from transformers import AutoModelForCausalLM
from datasets import load_dataset
from trl import SFTTrainer

dataset = load_dataset("imdb", split="train")

model = AutoModelForCausalLM.from_pretrained("facebook/opt-350m")

trainer = SFTTrainer(
    model,
    train_dataset=dataset,
    dataset_text_field="text",
)

trainer.train()
```

cc @lvwerra

### CODE REVIEW

File: trl/trainer/sft_trainer.py
Comment (lvwerra) [inline trl/trainer/sft_trainer.py:229]: I think you can just tokenize the dataset so you have the input ids. i would not pad at all (the collator will do this, and if a batch has elements that are all shorter it will be faster).

then you can just pass the tokenized dataset along the data collator (`data_collator = DataCollatorForLanguageModeling(tokenizer, mlm=False)`)

we did something similar here: https://huggingface.co/learn/nlp-course/chapter7/6?fw=pt

then we can maybe just call the function `_tokenize_dataset`

File: trl/trainer/sft_trainer.py
Comment (lvwerra) [inline trl/trainer/sft_trainer.py:162]: don't we need the same test for eval?

**Review by lvwerra — COMMENTED** [review]
Getting into great shape! Left a few comments.

File: trl/trainer/utils.py
Comment (younesbelkada) [reply trl/trainer/utils.py:234]: :D

File: trl/trainer/sft_trainer.py
Comment (younesbelkada) [reply trl/trainer/sft_trainer.py]: Yeah makes sense

File: trl/trainer/sft_trainer.py
Comment (younesbelkada) [reply trl/trainer/sft_trainer.py:113]: The `dataset_kwargs` corresponds to the kwargs of `ConstantLengthDataset`, there are 6 optional of them, I think we should move them to proper kwarg as we can always modify that class, but for `prepare_int8_training` kwargs I would maybe keep them as they are

File: trl/trainer/sft_trainer.py
Comment (younesbelkada) [reply trl/trainer/sft_trainer.py:252]: Thanks for the great pointer!

File: trl/trainer/sft_trainer.py
Comment (younesbelkada) [reply trl/trainer/sft_trainer.py]: I changed the way we create the datasets a bit!

File: trl/trainer/sft_trainer.py
Comment (younesbelkada) [reply trl/trainer/sft_trainer.py]: Or maybe we should educate users to create models outside the trainer in case they want to have full control over that function and remove `prepare_int8_training_kwargs`. Wdyt?

File: docs/source/sft_trainer.mdx
Comment (lewtun) [inline docs/source/sft_trainer.mdx:22]: nit:
```suggestion
You can also construct a model outside of the trainer and pass it as follows:
```

File: docs/source/sft_trainer.mdx
Comment (lewtun) [inline docs/source/sft_trainer.mdx:81]: Should these references to the classes include doc links?

File: docs/source/sft_trainer.mdx
Comment (lewtun) [inline docs/source/sft_trainer.mdx:134]: Have you tested this in distributed settings? I found that one usually needs to ensure the model is saved only on the main process or you cannot reload the model due to a `meta device` error.

File: docs/source/sft_trainer.mdx
Comment (lewtun) [inline docs/source/sft_trainer.mdx:141]: It would be nice to see how the callbacks should be added - maybe with a simple training loop?

File: docs/source/sft_trainer.mdx
Comment (lewtun) [inline docs/source/sft_trainer.mdx:147]:
```suggestion
- `SFTTrainer` always pads by default the sequences to the `max_seq_length` argument of the `SFTTrainer`. If none is passed, the trainer will retrieve that value from the tokenizer. Some tokenizers do not provide default value, so there is a check to retrieve the minimum between 2048 and that value. Make sure to check it before training.
```

File: docs/source/sft_trainer.mdx
Comment (lewtun) [inline docs/source/sft_trainer.mdx:150]:
```suggestion
- If you create a model outside the trainer, make sure to not pass to the trainer any additional keyword arguments that are relative to `from_pretrained()` method.
```

File: trl/trainer/sft_trainer.py
Comment (lewtun) [inline trl/trainer/sft_trainer.py:89]: For instruction fine-tuning, it is quite common to have two columns: one for the prompt & the other for the response. This allows people to format examples like Alpaca did as follows:

```
Below is an instruction ...

### Instruction
{prompt}

### Response:
{completion}
```

Perhaps it would be good to show users in the docs the two type of usage, ie a dataset pre-formatted and then one using the `formatting_func`?

File: trl/trainer/sft_trainer.py
Comment (lewtun) [inline trl/trainer/sft_trainer.py:100]: Does `trl` provide a helper method for this? If yes, it would be nice to see a small example in the docs of how this dataset works

**Review by lewtun — COMMENTED** [review]
Thanks for adding this sweet feature @younesbelkada 🔥 !

I've left a few questions and suggestions for things that could help improve user understanding. I'll let the core maintainer approve this :)

File: tests/test_sft_trainer.py
Comment (younesbelkada) [reply tests/test_sft_trainer.py:121]: Thanks ! Added a test in [1f86b70](https://github.com/lvwerra/trl/pull/323/commits/1f86b70893e0172044f6a6e8168994238b0189fb)

File: docs/source/sft_trainer.mdx
Comment (younesbelkada) [reply docs/source/sft_trainer.mdx:138]: Since the trainer is not integrated with `accelerate` I can't really test it out of the box in distrubuted setting it seems.. 
however looking at the trainer source code, it adds a lot of checks using `args.shoud_save`, I believe this automatically takes care of the distrubted setting: https://github.com/huggingface/transformers/blob/b8648290d2d97e7c7dbccd2d4a6a4f44e70d3b63/src/transformers/training_args.py#L1747

File: docs/source/sft_trainer.mdx
Comment (younesbelkada) [reply docs/source/sft_trainer.mdx:145]: Awesome! Added it here: [6097010](https://github.com/lvwerra/trl/pull/323/commits/609701062ca2bcf4a59410edecf0cbde7e391e7f)

File: trl/trainer/sft_trainer.py
Comment (younesbelkada) [reply trl/trainer/sft_trainer.py:90]: Yes makes sense, added it in dff614c4c37679c180c72a4401598d64c1514ecf

File: docs/source/sft_trainer.mdx
Comment (younesbelkada) [reply docs/source/sft_trainer.mdx]: The references should be now added!

File: trl/trainer/sft_trainer.py
Comment (younesbelkada) [reply trl/trainer/sft_trainer.py:101]: I added a pointer to an example that uses this in 91b2643160cbf199558fb1f8cc5946c1c2f09e1e

File: docs/source/sft_trainer.mdx
Comment (lvwerra) [inline docs/source/sft_trainer.mdx:3]:
```suggestion
Supervised fine-tuning (or SFT for short) is a crucial step in RLHF. In TRL we provide an easy-to-use API to create your SFT models and train them with few lines of code on your dataset.
```

File: trl/trainer/sft_trainer.py
Comment (lvwerra) [inline trl/trainer/sft_trainer.py:131]: I am slightly opposed to pass kwargs like that. I think if people want to use something other than default they should just load the model outside. It's just one additional line of code for them. We on the other hand need to then worry about those: e.g. if someone has a typo in one of the above kwargs they will get a weird error because it's passed to the model.

What do you think?

File: trl/trainer/sft_trainer.py
Comment (lvwerra) [inline trl/trainer/sft_trainer.py:266]: this will load the full dataset into memory. i think we should rather pass the column name to the `ConstantLengthDataset`. In the CLD: if a column name is provided then `self.formatting_func = lambda x: x[column_name]` and else the prerpocessing function. What do you think?

File: trl/trainer/sft_trainer.py
Comment (lvwerra) [inline trl/trainer/sft_trainer.py:293]: I think we should probably set this to False. If you are not using packing then having batches with have samples is a bit odd. What do you think?

File: trl/trainer/sft_trainer.py
Comment (lvwerra) [inline trl/trainer/sft_trainer.py:178]: haha why 2048? some models have more context window. also then we should probably emit a warning if we are not using the tokenizer's maximum.

File: trl/trainer/sft_trainer.py
Comment (younesbelkada) [reply trl/trainer/sft_trainer.py:131]: Yeah totally aligned on this!

File: trl/trainer/sft_trainer.py
Comment (younesbelkada) [reply trl/trainer/sft_trainer.py:289]: Yes makes sense!

File: trl/trainer/sft_trainer.py
Comment (younesbelkada) [reply trl/trainer/sft_trainer.py]: ahaha

Comment (younesbelkada) [reply trl/trainer/sft_trainer.py]: Fixed the maximum to 496 and added a warning message in [710c998](https://github.com/lvwerra/trl/pull/323/commits/710c99879860e5d02afb8dea9ba50b386f6959d4)

File: trl/trainer/sft_trainer.py
Comment (younesbelkada) [reply trl/trainer/sft_trainer.py:266]: Sounds great!

**Review by lvwerra — APPROVED** [review]

**Comment by lvwerra** [comment]
No, the SFTTrainer only trains on the text with the causal language modeling objective.

**Comment by hezhiyang2000** [comment]
Is this function suitable to train a instruction-following LLM? I reviewed the code and can't find the code about labels that avoiding to calculate the loss of instruction, prompts and labels.

**Comment by lvwerra** [comment]
See the docs here: https://huggingface.co/docs/trl/sft_trainer#train-on-completions-only

**Pull request merged**
